test_cases/stdlib/itertools/check_itertools_recipes.py

```diff
@@ -0,0 +1,368 @@
+"""Type-annotated versions of the recipes from the itertools docs.
+
+These are all meant to be examples of idiomatic itertools usage,
+so they should all type-check without error.
+"""
+from __future__ import annotations
+
+import collections
+import math
+import operator
+import sys
+from itertools import chain, combinations, count, cycle, filterfalse, islice, repeat, starmap, tee, zip_longest
+from typing import Any, Callable, Hashable, Iterable, Iterator, Sequence, TypeVar, overload
+from typing_extensions import Literal, TypeVarTuple, Unpack
+
+_T = TypeVar("_T")
+_T1 = TypeVar("_T1")
+_T2 = TypeVar("_T2")
+_HashableT = TypeVar("_HashableT", bound=Hashable)
+_Ts = TypeVarTuple("_Ts")
+
+
+def take(n: int, iterable: Iterable[_T]) -> list[_T]:
+    "Return first n items of the iterable as a list"
+    return list(islice(iterable, n))
+
+
+# Note: the itertools docs uses the parameter name "iterator",
+# but the function actually accepts any iterable
+# as its second argument
+def prepend(value: _T1, iterator: Iterable[_T2]) -> chain[_T1 | _T2]:
+    "Prepend a single value in front of an iterator"
+    # prepend(1, [2, 3, 4]) --> 1 2 3 4
+    return chain([value], iterator)
+
+
+def tabulate(function: Callable[[int], _T], start: int = 0) -> Iterator[_T]:
+    "Return function(0), function(1), ..."
+    return map(function, count(start))
+
+
+def repeatfunc(func: Callable[[Unpack[_Ts]], _T], times: int | None = None, *args: Unpack[_Ts]) -> Iterator[_T]:
+    """Repeat calls to func with specified arguments.
+
+    Example:  repeatfunc(random.random)
+    """
+    if times is None:
+        return starmap(func, repeat(args))
+    return starmap(func, repeat(args, times))
+
+
+def flatten(list_of_lists: Iterable[Iterable[_T]]) -> chain[_T]:
+    "Flatten one level of nesting"
+    return chain.from_iterable(list_of_lists)
+
+
+def ncycles(iterable: Iterable[_T], n: int) -> Iterator[_T]:
```

Comment:
```suggestion
def ncycles(iterable: Iterable[_T], n: int) -> chain[_T]:
```

Reply:
I think an Iterator return is actually better here; there's nothing useful you can do with a chain object you can't do with any Iterator, and it's basically an implementation detail exactly what kind of iterator you're returning.

Reply:
I think there's exactly one useful thing that you can do with a `chain` that you can't do with another other `Iterator`: you can access the `from_iterable` classmethod.

Accessing an alternative constructor like that from an instance is pretty unusual, though, and I'm inclined to agree that the precise kind of iterator being returned here is probably an implementation detail. I think there's another place in this file where I already have `-> chain`; I'm actually inclined to change _that_ one so that it's `-> Iterator`

Reply:
> I think an Iterator return is actually better here; there's nothing useful you can do with a chain object you can't do with any Iterator, and it's basically an implementation detail exactly what kind of iterator you're returning.

In that case, we should change the annotations that are returning `chain` to `Iterator` as well.